spec/ssh_scan/rule_engine_spec.rb

```diff
@@ -0,0 +1,125 @@
+
+    it "should ignore attributes not in results" do
+      result = {:ssh_version => 2.0}
+      expect(SSHScan::RuleEngine.matchSingleRule("ssh_lib_eq", "openssh", result)).to be true
+    end
+  end
+
+  # Note: inputs can be given in a hash as well
+  context "parse conditional boolean expressions" do
+    it "simple true/false" do
+      expect(SSHScan::RuleEngine.parseBooleanRule("x", "x" => true)).to be true
+      expect(SSHScan::RuleEngine.parseBooleanRule("x", "x" => false)).to be false
+    end
+
+    it "should work for NOT expressions" do
+      expect(SSHScan::RuleEngine.parseBooleanRule("!x", "x" => true)).to be false
+      expect(SSHScan::RuleEngine.parseBooleanRule("!x", "x" => false)).to be true
+    end
+
+    it "should work for AND expressions" do
+      expect(SSHScan::RuleEngine.parseBooleanRule("x && y", "x" => true, "y" => true)).to be true
+      expect(SSHScan::RuleEngine.parseBooleanRule("x && y", "x" => true, "y" => false)).to be false
+      expect(SSHScan::RuleEngine.parseBooleanRule("x && y", "x" => false, "y" => true)).to be false
+      expect(SSHScan::RuleEngine.parseBooleanRule("x && y", "x" => false, "y" => false)).to be false
+    end
+
+    it "should work for OR expressions" do
+      expect(SSHScan::RuleEngine.parseBooleanRule("x || y", "x" => true, "y" => true)).to be true
+      expect(SSHScan::RuleEngine.parseBooleanRule("x || y", "x" => true, "y" => false)).to be true
+      expect(SSHScan::RuleEngine.parseBooleanRule("x || y", "x" => false, "y" => true)).to be true
+      expect(SSHScan::RuleEngine.parseBooleanRule("x || y", "x" => false, "y" => false)).to be false
+    end
+
+    it "should work for compounded non bracketed AND, OR, NOT expressions" do
+      expect(SSHScan::RuleEngine.parseBooleanRule("!x && y", "x" => false, "y" => true)).to be true
+      expect(SSHScan::RuleEngine.parseBooleanRule("x || !y", "x" => false, "y" => false)).to be true
+    end
+
+    it "should work for compounded bracketed AND, OR, NOT expressions" do
+      expect(SSHScan::RuleEngine.parseBooleanRule("x && (y || z)",
+        "x" => false, "y" => true, "z" => false)).to be false
+      expect(SSHScan::RuleEngine.parseBooleanRule("!x || (y && (z || (!var)))",
+        "x" => true, "y" => true, "z" => true, "var" => true)).to be true
+    end
+  end
+
+  context "match simple rules + ruleset condition" do
+    it "should be able to match complex conditions with OR, AND, NOT" do
+      result1 = {:ssh_version => 2.0, :ssh_lib => "openssh"}
+      result2 = {:ssh_version => 1.99, :ssh_lib => "dropbear"}
+      result3 = {:ssh_version => 2.1, :ssh_lib => "dropbear"}
+      rules = {
+        "a" => ["ssh_version_gt", 1.99],
+        "b" => ["ssh_lib_eq", "openssh"],
+        "c" => ["ssh_lib_eq", "dropbear"],
+        "d" => ["ssh_version_lt", 1.99]
+      }
+      ruleset = "(a && b) || (c && !d)" # !d means ssh_version_gte 1.99
+      expect(SSHScan::RuleEngine.matchRuleset(rules, ruleset, result1)).to be true
```

Comment:
Walking through the logic out loud...

- ssh version 2.0 is gt than 1.99 (OK)
- ssh lib openssh is eq openssh (OK)

Overall: OK